drivers/adc/adc_stm32.c

```diff
@@ -758,71 +772,78 @@ static int start_read(const struct device *dev,
+		 * Only for ADC1 instance (ADC4 has no Channel preselection capability).
+		 */
+		if (adc == ADC1) {
+			LL_ADC_SetChannelPreselection(adc, channel);
+		}
 #endif
+
 #if defined(CONFIG_SOC_SERIES_STM32F0X) || \
 	defined(CONFIG_SOC_SERIES_STM32L0X)
-	LL_ADC_REG_SetSequencerChannels(adc, channel);
+		LL_ADC_REG_SetSequencerChannels(adc, channel);
 #elif defined(CONFIG_SOC_SERIES_STM32WLX)
-	/* Init the the ADC group for REGULAR conversion*/
-	LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_CONFIGURABLE);
-	LL_ADC_REG_SetTriggerSource(adc, LL_ADC_REG_TRIG_SOFTWARE);
-	LL_ADC_REG_SetSequencerLength(adc, LL_ADC_REG_SEQ_SCAN_DISABLE);
-	LL_ADC_REG_SetOverrun(adc, LL_ADC_REG_OVR_DATA_OVERWRITTEN);
-	LL_ADC_REG_SetSequencerRanks(adc, LL_ADC_REG_RANK_1, channel);
-	LL_ADC_REG_SetSequencerChannels(adc, channel);
-	/* Wait for config complete config is ready */
-	while (LL_ADC_IsActiveFlag_CCRDY(adc) == 0) {
-	}
-	LL_ADC_ClearFlag_CCRDY(adc);
+		/* Init the the ADC group for REGULAR conversion*/
+		LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_CONFIGURABLE);
+		LL_ADC_REG_SetTriggerSource(adc, LL_ADC_REG_TRIG_SOFTWARE);
+		LL_ADC_REG_SetSequencerLength(adc, LL_ADC_REG_SEQ_SCAN_DISABLE);
+		LL_ADC_REG_SetOverrun(adc, LL_ADC_REG_OVR_DATA_OVERWRITTEN);
+		LL_ADC_REG_SetSequencerRanks(adc, LL_ADC_REG_RANK_1, channel);
+		LL_ADC_REG_SetSequencerChannels(adc, channel);
+		/* Wait for config complete config is ready */
+		while (LL_ADC_IsActiveFlag_CCRDY(adc) == 0) {
+		}
+		LL_ADC_ClearFlag_CCRDY(adc);
 #elif defined(CONFIG_SOC_SERIES_STM32G0X)
-	/* STM32G0 in "not fully configurable" sequencer mode */
-	LL_ADC_REG_SetSequencerChannels(adc, channel);
-	LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_FIXED);
-	while (LL_ADC_IsActiveFlag_CCRDY(adc) == 0) {
-	}
-	LL_ADC_ClearFlag_CCRDY(adc);
-#elif defined(CONFIG_SOC_SERIES_STM32U5X)
-	if (adc != ADC4) {
-		LL_ADC_REG_SetSequencerRanks(adc, table_rank[0], channel);
-		LL_ADC_REG_SetSequencerLength(adc, table_seq_len[0]);
-	} else {
+		/* STM32G0 in "not fully configurable" sequencer mode */
+		LL_ADC_REG_SetSequencerChannels(adc, channel);
 		LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_FIXED);
-		LL_ADC_REG_SetSequencerLength(adc,
-					      BIT(__LL_ADC_CHANNEL_TO_DECIMAL_NB(channel)));
-	}
+		while (LL_ADC_IsActiveFlag_CCRDY(adc) == 0) {
+		}
+		LL_ADC_ClearFlag_CCRDY(adc);
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+		if (adc != ADC4) {
+			LL_ADC_REG_SetSequencerRanks(adc, table_rank[0], channel);
+			LL_ADC_REG_SetSequencerLength(adc, table_seq_len[0]);
+		} else {
+			LL_ADC_REG_SetSequencerConfigurable(adc, LL_ADC_REG_SEQ_FIXED);
+			LL_ADC_REG_SetSequencerLength(adc,
+						      BIT(__LL_ADC_CHANNEL_TO_DECIMAL_NB(channel)));
+		}
 #else
-	LL_ADC_REG_SetSequencerRanks(adc, table_rank[0], channel);
-	LL_ADC_REG_SetSequencerLength(adc, table_seq_len[0]);
+		LL_ADC_REG_SetSequencerRanks(adc, table_rank[data->channel_count], channel);
```

Comment:
`channel_count` is almost correct.  The ordinal position of the channel in the sequence should be used here.  `data->channels` is a bit mask, so the first channel in the sequence is not necessarily channel 0, and so on.

But I think it's off by 1.  The first channel should use table_rank[0] to be programmed as the first sequencer slot.  But data->channel_count is incremented at the start of loop, so it will be 1.

OTOH, it looks like `table_seq_len[0]` means zero channels in the sequence, so the correct index in the table is being used for that table.

Reply:
Indeed, it was off by one. I missed this change because the `data->channels++` was originally done at the end of the loop. And this applies to both `rank` and `seq_len`, because with a length of 1 the sequencer should be disabled. 

I changed how it's calculated to be easier to read by using a `channel_index` which is incremented at the end of the loop.

``` C
		LL_ADC_REG_SetSequencerRanks(adc, table_rank[channel_index], channel);
		LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
```

Reply:
Without LTO, using local variables as loop indexes is better anyway.  The compiler has a hard time determining that no function calls will access `data->channels` when it can only see this one file.  So it must update the value in RAM.  But the compiler can see that no external functions can access the local variable so it can just leave it in a register.